extensions/graphql/src/graphql.server.ts

```diff
@@ -148,28 +161,61 @@ export class GraphQLServer extends Context implements Server {
 
   async start() {
     const schema = await this._setupSchema();
-
+    this.schema = schema;
     // Allow a graphql context resolver to be bound to GRAPHQL_CONTEXT_RESOLVER
-    const graphqlContextResolver: ContextFunction<ExpressContext> =
+    const graphqlContextResolver: ExpressMiddlewareOptions<BaseContext>['context'] =
       (await this.get(GraphQLBindings.GRAPHQL_CONTEXT_RESOLVER, {
         optional: true,
-      })) ?? (context => context);
+      })) ?? (async context => context);
+    // Allow a graphql context resolver to be bound to GRAPHQL_CONTEXT_RESOLVER
+    const graphqlWsContextResolver =
+      (await this.get(GraphQLBindings.GRAPHQL_WS_CONTEXT_RESOLVER, {
+        optional: true,
+      })) ??
+      (async (context: object) => {
+        return context;
+      });
 
     // Create ApolloServerExpress GraphQL server
-    const serverConfig: ApolloServerExpressConfig = {
-      context: graphqlContextResolver,
+    const serverConfig = {
       ...this.options.apollo,
       schema,
-    };
+      status400ForVariableCoercionErrors: true,
+      // plugins: [ApolloServerPluginDrainHttpServer({ httpServer: this.httpServer })],
+    } as ApolloServerOptionsWithSchema<BaseContext>;
     const graphQLServer = new ApolloServer(serverConfig);
     await graphQLServer.start();
-    graphQLServer.applyMiddleware({
-      app: this.expressApp,
-      ...this.options.middlewareOptions,
-    });
-
+    this.expressApp.use(
+      this.options.graphQLPath ?? '/',
+      cors<cors.CorsRequest>(),
+      json(),
+      expressMiddleware(graphQLServer, {
+        context: graphqlContextResolver,
+        ...this.options.middlewareOptions,
+      }),
+    );
     // Start the http server if created
     await this.httpServer?.start();
+
+    let server;
+    if (this.options.asMiddlewareOnly) {
+      const rest: RestServer = await this.get('servers.RestServer');
+      server = rest.httpServer?.server;
+    } else {
+      server = this.httpServer?.server;
+    }
```

Comment:
Could we spin this out GraphQL-over-WebSocket its own package?

`@loopback/rest` should be marked as a peer dependency due to `asMiddlewareOnly`-mode, which would bloat the core graphql package quite a bit.

The current setup is also a little fragile as it assumes `@loopback/rest` is already bootstrapped in the application.

```suggestion
    if (this.options.asMiddlewareOnly) {
      const rest: RestServer = await this.get(RestBindings.RestServer);
      server = rest.httpServer?.server;
    } else {
      server = this.httpServer?.server;
    }
```

Reply:
> Could we spin this out GraphQL-over-WebSocket its own package?

GraphQL subscriptions require a transport mechanism other than HTTP, such as WebSockets, Server-Sent Events, or HTTP Multipart. My PR addresses the issue with GraphQL subscriptions, which are currently not functioning in the master branch. I opted to implement subscriptions over WebSocket using graphql-ws because it is the most popular solution.
If you have any ideas on how we might move this into a separate package, I would greatly appreciate your suggestion.